agent/session.go

```diff
@@ -67,6 +67,10 @@ func newSession(ctx context.Context, agent *Agent, delay time.Duration, sessionI
 	)
 
 	if err != nil {
+		// since we are returning without launching the session goroutine, we
+		// need to provide the delay that is guaranteed by calling this
+		// function.
+		time.Sleep(delay)
```

Comment:
Why sleep instead of a timer ? 
https://github.com/docker/swarmkit/pull/2570/files#diff-15ffe95a2da45f70696ffa3c01949601L84

Reply:
Also, would this possibly block here:  https://github.com/docker/swarmkit/blob/master/agent/agent.go#L367, thus preventing any other events, such as shutdown or node update, etc. from happening?

Not sure if it would make sense instead to do the sleep/write the error to the channel in a goroutine instead?

Reply:
Yes, this probably isn't a great fix. For this to be correct, we need a goroutine. Good catch!